src/main/java/bio/terra/buffer/service/resource/projectid/GcpProjectIdGenerator.java

```
@@ -22,10 +30,58 @@
   /** The largest number to use for random suffixes for adjective/noun project IDs. */
   @VisibleForTesting static final int RANDOM_SUFFIX_LIMIT = 10000;
 
+  /** The maximum allowed length for a GCP project id. */
+  private static final int MAX_LENGTH_GCP_PROJECT_ID = 30;
+
+  /** The maximum allowed length for a GCP project id prefix (e.g. "terra-"). */
+  public static final int MAX_LENGTH_GCP_PROJECT_ID_PREFIX = 12;
+
+  /**
+   * The maximum number of times to try generating a project id that has an allowable length and is
+   * not already in use.
+   */
+  private static final int MAX_RETRIES = 100;
+
+  /**
+   * Generate a GCP project id from the prefix and scheme.
+   *
+   * <p>Keep retrying the id generation until we find one that fits into the GCP project id length
+   * limit and is not already in use, or until we have retried the maximum number of times.
+   *
+   * <p>To check if a project is in use, this method tries to retrieve the project. This will only
+   * succeed if RBS has permission to get that project. We expect this to be the case most of the
+   * time because of the Terra-specific prefix, but it's possible this step will generate a project
+   * that is in use outside of Terra. In that case, the project creation step will fail.
+   *
+   * @param projectIdSchema prefix and naming scheme
+   * @param rmCow resource manager wrapper to retrieve the project
+   * @return generated project id (prefix + naming scheme)
+   */
+  public String generateIdWithRetries(
+      ProjectIdSchema projectIdSchema, CloudResourceManagerCow rmCow) {
+    for (int numTries = 0; numTries < MAX_RETRIES; numTries++) {
+      String projectId = generateId(projectIdSchema);
+      if (projectId.length() <= MAX_LENGTH_GCP_PROJECT_ID)
+        try {
+          if (retrieveProject(rmCow, projectId).isEmpty()) {
```

this pattern may not work well in limited GCP API quota environment. `retrieveProject` is a google call, and we expect GCP calls' retry interval is slightly longer because those quota is x numbers/**minute** 
That's why we have [newCloudApiDefaultRetryRule](https://github.com/DataBiosphere/terra-resource-buffer/blob/af480aeda9f0dd336065536daa6241b78b7123ec/src/main/java/bio/terra/buffer/service/resource/flight/StepUtils.java#L20) and [newInternalDefaultRetryRule](https://github.com/DataBiosphere/terra-resource-buffer/blob/af480aeda9f0dd336065536daa6241b78b7123ec/src/main/java/bio/terra/buffer/service/resource/flight/StepUtils.java#L28)

Suggest to split generateId and retrieveProject like what we are doing now, 

this method here just retry this method with no sleep:
```
 if (projectId.length() <= MAX_LENGTH_GCP_PROJECT_ID) {
   return projectId;
} 
```

In Flight step: 
```
try {
generateStep()
retrieveProject()
} catch (IOE | RuntimeE) {
return step retryable error;
}
```

One more note is in the flight step which calls this method, shall we catch exception that throwed by this method and return retry-able failure?  If just throw RuntimeException, Stairway may just fail that flight with no retries: 
https://github.com/DataBiosphere/stairway/blob/develop/stairway/src/main/java/bio/terra/stairway/Flight.java#L301

---

Good point about potential quota issues. Two things I'm not sure about:
- In this case, we're not retrying the `retrieveProject` call, we're retrying the project id generation. So I think it only makes sense to use Stairway retries for sleeping after an `IOException`. If we find an existing project with this name, then we can just try to generate a new id immediately.
- I don't think we need to use Stairway retries for calls to `generateId`. This just does a random word lookup, so sleeping won't help anything. I think we can just fail the flight at that point.

---

1: What you mean by "we're not retrying the retrieveProject call, we're retrying the project id generation." I think you do retry `retrieveProject` call when the previous project id is already being used. Also the quota issues happens when we submit large number of flights at the same time. For example, creating a new pool, and that require create large number of CreationFlight at one time. There will be many retrieveProject call fired by other same steps.
2: In your code, your throw RuntimeException. That one is not captuered by Stairway. Would Stairway just treat this as a failure flight with no retry? I think so. But I don't think it is a 'retry-able' Stairway step.

---

The suggestion above is not to make all retries in Stairway retried. it make sense to retry immediately if project id too long. 
Then the  `GenerateProjectIdStep` get the short project id back from the static method, it use that project id to make a retrieve call, that is part we Stairway retry can be used.

---

1: If `retrieveProject` finds that project id 123 is in use, then we generate a new project id 456 and call `retrieveProject` on that. It's not the same call that's being retried over and over. That's why I'm not sure sleeping between calls makes sense. e.g. if `retrieveProject(123)` finds a project, then sleeping 10 minutes before a call to `retrieveProject(456)` shouldn't affect the outcome.  That's why I didn't include a sleep (via stairway retries) between calls to `retreiveProject`.

However, if `retrieveProject` fails with an `IOException` then that may be a quota issue as you pointed out. So I agree we should sleep (via stairway retries) before the next call. Happy to discuss on GVC if that's better.

2: Yes, stairway would just fail this flight with no retry. Would it be better if I made that explicit? (e.g. throw a checked exception and then return a FATAL status in the step)

---

1: `if retrieveProject(123) finds a project, then sleeping 10 minutes before a call to retrieveProject(456)` where is that logic happened? 
My understanding of your code: 
```
for (int numTries = 0; numTries < MAX_RETRIES; numTries++) {
      String projectId = generateId(projectIdSchema);
      if (projectId.length() <= MAX_LENGTH_GCP_PROJECT_ID)
        if (retrieveProject(rmCow, projectId).isEmpty()) {
          logger.info("Generated GCP project id after {} tries.", numTries);
          return projectId;
        } else {
          logger.info("Generated GCP project is already in use: {}", projectId);
        }
    }
```
The sequence can be: 
 a: generate 123
b: retrieve 123, used
c: generate 456
d: retrieve 456, also used,
e, f, g....

I did not find the 10 min sleep during b and d. 

2: I am fine with what you have now: Not catching anything and just let Stairway failed with no retry. Initially I throught catch exception then return **a retryable status in the step** because it is retryable IMO. But I think what you have is also fine, just let it fail, and a new flight will be kicked off, and do the same thing..

---

Sorry, if my comment is confusing. My suggestion is similar to the the code in my previous comment:
https://github.com/DataBiosphere/terra-resource-buffer/pull/180#discussion_r681788451 pull retrieveProject() out from the helper method : `generateId()`, and make it back to project creation step. Your helper generateId(). Let me make a PR to explain it

---

There it is: https://github.com/DataBiosphere/terra-resource-buffer/pull/181/files 
What you said: `I move retrieveProject into the step and return a retry exception if it failed` I did not see that is happening. retrieveProject is not in the step, it is still in generateId() method with no retry

---

I don't want to block your PR. Given the 12 char limit you add, and the low chance of having conflict, no sleep between  `retrieveProject` is not my big concern for now. Totally fine to proceed with your change. 
LGTM

---

OK to summarize, I think the two options we're talking about are:
option A: retrieve project in generate project id step, and try to generate another one if it's already used.
option B: retrieve the project only in the create project step, and fail the flight if it's already used.

From your earlier comment, I think these are similar because if a flight fails then we will just kick off a new flight to fill the pool. The main difference is that option A retries within a step and option B retries the whole flight. This means the time between attempts to generate an unused project id will be different (zero for option A, however long it takes to start a new flight for option B).

I'm going to put some time on your calendar to make sure we're on the same page.

---

Talked to @yonghaoy and we decided on the following:
- Use stairway retry if the generated project id is already found. This is so that we know each stairway step execution only makes a single get call, which makes it easier to decide how many flights can run in parallel. We could immediately try generating a new project id and checking if it exists, but that would mean each stairway step execution could make a variable number of get calls, which would make the above calculation harder.
- Move the `retreiveProject` call from the `CreateProjectStep` to the `GenerateProjectIdStep`, so that we lower the risk of the undo failing because of a project get quota issue. The generate project id step undo is a no-op, because no project has been created yet. The create project step undo includes another project get call so we can try to delete it.